// @ant-design/bisheng-plugin/lib/VideoPlayer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = VideoPlayer;

var _react = _interopRequireDefault(require("react"));

var _reactSublimeVideo = _interopRequireDefault(require("react-sublime-video"));

function VideoPlayer(_ref) {
  var video = _ref.video;
  var alt = video.alt,
      description = video.description,
      src = video.src;
  var videoClassName = video["class"];
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "preview-image-box ".concat(videoClassName)
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: 'preview-image-wrapper'
  }, /*#__PURE__*/_react["default"].createElement(_reactSublimeVideo["default"], {
    src: src,
    type: "video/mp4",
    loop: true
  })), /*#__PURE__*/_react["default"].createElement("div", {
    className: "preview-image-title"
  }, alt), /*#__PURE__*/_react["default"].createElement("div", {
    className: "preview-image-description",
    dangerouslySetInnerHTML: {
      __html: description
    }
  }));
}
